core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -640,26 +617,22 @@ public boolean equals(Object o) {
         if (this == o) {
             return true;
         }
-        // we check on instanceof so we also handle the ImmutableShardRouting case as well
         if (o == null || !(o instanceof ShardRouting)) {
             return false;
         }
         ShardRouting that = (ShardRouting) o;
-        if (version != that.version) {
-            return false;
-        }
         if (unassignedInfo != null ? !unassignedInfo.equals(that.unassignedInfo) : that.unassignedInfo != null) {
             return false;
         }
         return equalsIgnoringMetaData(that);
     }
 
-    private long hashVersion = version - 1;
+    private boolean usePreComputedHashCode = false;
     private int hashCode = 0;
 
     @Override
     public int hashCode() {
-        if (hashVersion == version) {
+        if (frozen && usePreComputedHashCode) {
```

Comment:
ok, some clarification here (as the story is way more complicated... :-/ ). Precomputing the hash in the old version was actually broken, it would never cache (could as well have put `assert hashVersion != version` there). I fixed caching in this PR for the case where we are frozen (simplest change). I can extend the caching to non-frozen shard routings as well.

Now, looking at the usage sites of this (did a grep for `Map<ShardRouting`), there is one where non-frozen ShardRouting is used as map key and that is in BalancedShardsAllocator. This uncovers an issue in BalancedShardsAllocator: Using as map key a mutable object, which we modify in the same class... I wonder why this has not bitten us yet?